### DISCUSSION TITLE
Constraint on the effects that are out of the monadic flow.

### DISCUSSION THREAD

**Discussion by wclr** [haskell-effectful/effectful#141] [Q&A]
I would like to model a sample app here, one of the functions of it is to be able handle the request and return the response. In this model, it uses the effectful method `onRequest` to which we pass the handler (response returned will be converted to IO by interpreter and run on each request).

The problem I see here is that the `app` function in this case doesn't require `LogEff` effect constraint, which is used in the handler but is not used in the application monadic flow directly, it is only passed to handler here.

In my view, the `app's` signature doesn't reflect what the app is doing as a whole in this case. It reflects what the monadic flow is doing, not the app. What would you say about such an issue? Maybe it is possible somehow to achieve what I would like to see in this case?

```haskell
module EffectfulApp where

import Effectful (Eff, Effect, type (:>))
import Effectful.TH (makeEffect)


data LogEff :: Effect where
  LogLine :: String -> LogEff m ()
makeEffect ''LogEff


data Request = Request {reqUrl :: String}
data Response = Response {resText :: String}


type RequestHandler es =
  ( (LogEff :> es) => Request -> Eff es Response
  )


data ReqEff :: Effect where
  OnRequest :: (forall es. RequestHandler es) -> ReqEff m ()
makeEffect ''ReqEff


app ::
  ( ReqEff :> es
  -- LogEff : es -- is not required, and gives the waning
  ) =>
  Eff es ()
app = do
  onRequest \(Request {reqUrl}) -> do
    logLine $ "Got request from " <> reqUrl
    pure $ Response {resText = "Hello!"}

```

**Comment by wclr** [comment]
Well, I thought about it, maybe I'm not correct her, LogEff is present inside the ReqEff, so probably it is quite ok then.

**Comment by wclr** [comment]
Then I will ask here another question, related at least to this sample app.

Here is the interpretation and run module"
```haskell
module EffectfulAppEntry where

import EffectfulApp (LogEff (..), ReqEff (..), Request (..), app)

--
import Data.Function ((&))
import Effectful (
  Eff,
  IOE,
  MonadIO (..),
  runEff,
  type (:>),
 )
import Effectful.Dispatch.Dynamic (interpret)


runLog ::
  (IOE :> es) =>
  Eff (LogEff : es) a ->
  Eff es a
runLog = interpret \_ -> \case
  LogLine msg -> liftIO $ putStrLn msg


runReq ::
  (IOE :> es) =>
  Eff (ReqEff : es) a ->
  Eff es a
runReq = interpret $ const \case
  OnRequest toResponse ->
    liftIO $ do
      _ <-
        runEff
          ( toResponse Request {reqUrl = "/some"}
              & runLog
          )
      pure ()


main :: IO ()
main = do
  runEff $
    app
      & runReq
      -- we actually don't need logEff here but is is still can be applied
      & runLog
```

Why is it still possible to apply `logEff` interpretator to the program even if doesn't need it, and it can be just removed?

**Comment by eddiemundo** [comment]
The `es` in `Eff es a` is polymorphic and so the `runX`s determine what the concrete effect list is. If you have:
```haskell
app :: (Req :> es) => Eff es a
app = reqItUp

blah =
  app
    & runReq
    & runLogger
```
then the type of `blah` is something like `Eff [Req, Logger] a` and `[Req, Logger]` satisfies the constraint `Logger :> es`.

Doing the extra `logEff` shouldn't do anything because there are no log operations to interpret.

The interpreters determining the concrete stack is very similar to how mtl-style works.

**Comment by wclr** [comment]
I understand that it does nothing, but it seems to be misleading that such thing is allowed. In mtl, as you mentioned it, we can not hae yet another `runSmthT` on stack if it is not appropriate/needed, right?

**Comment by ocharles** [comment]
If you want your application to have a closed set of effects, then you need to specify that. In

```haskell
app :: (Req :> es) => Eff es a
```

You are saying `app` works for any `es`, provided `Req` is somewhere in `es`.

You could alternatively write

```haskell
app :: Eff '[Req] a
```

And now `app` can only be used with `runReq`.

That said, this isn't idiomatic. `runReq $ runLogger app` isn't really a problem to worry about, it's just a bit wasteful

**Comment by wclr** [comment]
Ok, I see it for the `app`, b I have a currenlty `runReq` signature as `(IOE :> es) => Eff (ReqEff : es) a -> Eff es a` and it doesn't work with closed `app`, 

I tried to replace it with `   Eff '[ReqEff] a -> Eff '[IOE] a` but get the error:
```
Couldn't match type: '[]
                 with: '[IOE]
  Expected: Eff '[ReqEff] a -> Eff '[IOE] a
    Actual: Eff '[ReqEff] a -> Eff '[] a
```
And can not get how to put the interperter correctly.

For my case the dummy interperter looks like:
```hs
runReq = interpret $ const \case
  OnRequest toResponse ->
    liftIO $ do
      _ <-
        runEff
          ( toResponse Request {reqUrl = "/some"}
              & runLog
          )
      pure ()
```

**Comment by ocharles** [comment]
I think we'd need to see full code that we can compile to help here, but I don't think you're going down a very productive path. This is not how `effectful` is designed to be used - keeping `es` polymorphic and adding constraints is the idiomatic approach and I think it would be best to try and help building understanding there.

**Comment by arybczak** [comment]
Indeed, what @ocharles said is accurate.

Using monomorphic effect stacks isn't even supported with higher order effects (see [here](https://hackage.haskell.org/package/effectful-core-2.2.2.1/docs/Effectful-Dispatch-Dynamic.html#t:SharedSuffix), specifically the last example), so you shouldn't really do things this way.

Btw, I'll convert this thread to a discussion, since it's not an issue.

**Comment by wclr** [comment]
@ocharles 
> I think we'd need to see full code that we can compile.

Well, there is a full code of modules that compiles in the first and the third post: modules `EffectfulApp` (logic) and `EffectfulAppEntry` (interpretation and run), besides the extensions used.

@arybczak 
> Using monomorphic effect stacks isn't even supported with higher order effects (see [here](https://hackage.haskell.org/package/effectful-core-2.2.2.1/docs/Effectful-Dispatch-Dynamic.html#t:SharedSuffix), specifically the last example), so you shouldn't really do things this way.

Could you elaborate more on what "monomorphic effect stacks" is and what is wrong with my sample's code/approach?

**Reply by arybczak** [reply]
Your original code is fine.

Monomorphic stack is just the first parameter of `Eff` that's, well, monomorphic, e.g. `Eff '[IOE] a`. `Eff es a` has a polymorphic stack that can be then constrained with `:>`, i.e. `IOE :> es => Eff es a`.

You can read more about this [here](https://hackage.haskell.org/package/effectful-core-2.2.2.1/docs/Effectful.html#g:8).

**Comment by wclr** [comment]
> This is not how effectful is designed to be used - keeping es polymorphic and adding constraints is the idiomatic approach and I think it would be best to try and help building understanding there.

So it is ok that, to attach interpreters that are not needed (won't be used) by the program at the moment? Is there a rationale behind this?

**Comment by ocharles** [comment]
Can you try and explain what the problem you see is? There is an almost insignificant performance hit and that's it

**Reply by wclr** [reply]
> Can you try and explain what the problem you see is? There is an almost insignificant performance hit and that's it

It is in my third post here in the comment:
```hs
main :: IO ()
main = do
  runEff $
    app
      & runReq
      -- we actually don't need runLog here but is is still can be applied
      & runLog
```

The problem I see that `runLog` is redundant here, but it is not shown anyhow. But maybe there is a rationale for this, so I'm asking about it.

**Reply by ocharles** [reply]
I get that, but you're not telling me what is _wrong_ with this. I can understand you don't want it, but I don't know _why_ you don't want it. It's a fairly rare situation to end up in - generally when chaining interpreters we are guided by the requirements of `app` - you first have `runEff app`, and then you're told you need a handler for `Req` so you add `runReq`, at which point the program compiles and we're done. Now `app` might once have required `Log` so you added `runLog`, but you later removed the `Log :> es` constraint. At this point `runLog` is redundant, but harmless. I'm not sure there's anything that `effectful` can reasonable do here (except maybe extending the plugin?) - we probably wouldn't want this to be an error, and it would be a lot of machinery to even get this to error.

**Reply by wclr** [reply]
Well, I'm probably just curious is this a feature with its ratinal reason or is this a limitation of type system? 

For example, as you described, we once had an effect then removed it and we don't need to change our interpreter chain in the code, at the same time we will need to change constraints in the signature, as it will show a warning about redundancy, which is obviously a good thing.

But in the case of interperter's chain there is no redundancy warnings, so is this a feature or is this just how type system works?

**Reply by wclr** [reply]
But you also pointed out that we could have `monomorphic` stack in the app's signature to make it possible to allow only needed interpreters in the chain.

But I could not modify my interpreters to compile with such a signature.

**Reply by ocharles** [reply]
I would say "how he type system works" is the best answer. It is probably possible to get what you want, but with a lot of work. In generally, Haskell constraints are somewhat like subtyping - we generally find instances that match a set of possible types, many of which will be larger than the smallest solution. This is not unique to `effectful`, `mtl` + `transformers` has a similar behavior - we can write:

```haskell
app :: Monad m => m ()
app = return ()

main :: IO ()
main = runReaderT (runReaderT app True) 'a'
```

Here `app` doesn't need anything, but it can be used with `runReaderT` (twice!).

**Reply by wclr** [reply]
Ok, but I would like to understand if it is possible to compile my code with closed effects stack, as it was mentioned in the beginning? The problem is that the `app` functions compiles, but I can not make an interpreter work with it.

**Reply by ocharles** [reply]
Could you share a full bit of Haskell code that we can try and compile?

**Reply by Kleidukos** [reply]
(and if you don't need an unreleased version of `effectful`, you can even share it on the playground: https://play-haskell.tomsmeding.com/)

**Reply by wclr** [reply]
This is the full code, I belive should work `app` has closed `Eff` type:

```hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE ExplicitNamespaces #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE RankNTypes  #-}


module EffectfulApp where

import Data.Function ((&))
import Effectful (Eff, Effect, type (:>))
import Effectful.TH (makeEffect)
import Effectful (
  Eff,
  IOE,
  MonadIO (..),
  runEff,
  type (:>),
 )
import Effectful.Dispatch.Dynamic (interpret)


data LogEff :: Effect where
  LogLine :: String -> LogEff m ()
makeEffect ''LogEff


data Request = Request {reqUrl :: String}
data Response = Response {resText :: String}


type RequestHandler es =
  ( (LogEff :> es) => Request -> Eff es Response
  )


data ReqEff :: Effect where
  OnRequest :: (forall es. RequestHandler es) -> ReqEff m ()
makeEffect ''ReqEff


app :: Eff '[ReqEff] ()
app = do
  onRequest \(Request {reqUrl}) -> do
    logLine $ "Got request from " <> reqUrl
    pure $ Response {resText = "Hello!"}

--


runLog ::
  (IOE :> es) =>
  Eff (LogEff : es) a ->
  Eff es a
runLog = interpret \_ -> \case
  LogLine msg -> liftIO $ putStrLn msg


runReq ::
  (IOE :> es) =>
  Eff (ReqEff : es) a ->
  Eff es a
runReq = interpret $ const \case
  OnRequest toResponse ->
    liftIO $ do
      _ <-
        runEff
          ( toResponse Request {reqUrl = "/some"}
              & runLog
          )
      pure ()


main :: IO ()
main = do
  runEff $
    app
      & runReq
      -- we actually don't need logEff here but is is still can be applied
      & runLog
```

**Reply by ocharles** [reply]
Well the problem is you are calling `runLog` right at the end, but `app`'s type is `Eff '[ReqEff]`, so `runLog` isn't compatible. But that's exactly what you want?

**Reply by wclr** [reply]
How my interpreters should be modified to make it work with closed effect types.

**Reply by ocharles** [reply]
@wclr I think the best I can offer is https://play-haskell.tomsmeding.com/saved/gQsOCcNP. The call to `inject` re-generalizes `app` to allow `IOE`

**Reply by wclr** [reply]
@ocharles Right, and then we go back to where we were in the beginning, now we can add redundant `runLog` again -)

**Reply by ocharles** [reply]
Yes. If you don't what that then you would need to annotate the type of `inject` - something like `inject @'[ReqEff] @'[ReqEff, IOE]`
